server/src/main/scala/com/cloudera/livy/server/batch/BatchSession.scala

```diff
@@ -23,62 +23,129 @@ import java.lang.ProcessBuilder.Redirect
 import scala.concurrent.{ExecutionContext, ExecutionContextExecutor}
 import scala.util.Random
 
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties
+
 import com.cloudera.livy.LivyConf
+import com.cloudera.livy.server.recovery.SessionStore
 import com.cloudera.livy.sessions.{Session, SessionState}
+import com.cloudera.livy.sessions.Session.RecoveryMetadata
+import com.cloudera.livy.util.LineBufferedProcess
 import com.cloudera.livy.utils.{AppInfo, SparkApp, SparkAppListener, SparkProcessBuilder}
 
+@JsonIgnoreProperties(ignoreUnknown = true)
+case class BatchRecoveryMetadata(
+    id: Int,
+    appTag: String,
+    appId: Option[String],
+    owner: String,
+    proxyUser: Option[String],
+    version: Int = 1)
+  extends RecoveryMetadata
+
+object BatchSession {
+  val RECOVERY_SESSION_TYPE = "batch"
+
+  def create(
+      id: Int,
+      owner: String,
+      proxyUser: Option[String],
+      livyConf: LivyConf,
+      createRequest: CreateBatchRequest,
+      sessionStore: SessionStore,
+      mockApp: Option[SparkApp] = None): BatchSession = {
+    val appTag = s"livy-batch-$id-${Random.alphanumeric.take(8).mkString}"
+    new BatchSession(
+      id, appTag, owner, proxyUser, livyConf, Left(createRequest), sessionStore, mockApp)
+  }
+
+  def recover(
+      m: BatchRecoveryMetadata,
+      livyConf: LivyConf,
+      sessionStore: SessionStore,
+      mockApp: Option[SparkApp] = None): BatchSession = {
+    new BatchSession(
+      m.id, m.appTag, m.owner, m.proxyUser, livyConf, Right(m.appId), sessionStore, mockApp)
+  }
+}
+
 class BatchSession(
```

Comment:
It's kind of a different situation.

In that patch, you were embedding code into the object that was passed down to the class instance as a parameter. So the class instance would do something, then call the code defined in the object, then do something else. That's confusing.

Here, you're passing extra parameters that are mutually exclusive, in a sense. For example, the `Either` parameter you had wasn't necessary; it was just being used as a way to move the `SparkApp` instantiation from the object into the `BatchSession` class. But `BatchSession`, ultimately, was only ever interested in the `SparkApp` instance, and not in how to build it. The `BatchSession` instance also doesn't care whether the `SparkApp` is a mock or not. So, it's better to move the logic for how to build that instance somewhere else - in this case, the factory methods which know how to do it.